File: TALLER_HILOS_NETBEANS_GR01/APLICATIVO/PARTE 3/Sincronizacion_Hilos_Banco/src/ec/edu/monster/modelo/Banco.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.monster.modelo;

import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 *
 * @author Maycol
 */

public class Banco {

    private final double[] cuentas;
    //LOCK para proporcionar capacidades de bloqueo en un contexto multihilo
    //ReeentrantLocck bloque que permite adquirir y liberar el bloqueo múltiples veces por el mismo hilo
    private Lock cierreBanco = new ReentrantLock();

    //Constructor inicializa 100 cuentas con saldo de 2000 dolares
    public Banco() {
        this.cuentas = new double[100];
        for (int i = 0; i < cuentas.length; i++) {
            cuentas[i] = 2000;
        }
    }
    
    //Se obtiene el saldo total de la cuenta
    public double getSaldoTotal() {
        double total = 0;
        for (double a : cuentas) {
            total += a;
        }
        return total;
    }
    
    public synchronized void transferencia(int cuentaOrigen, int cuentaDestino, double cantidad) throws InterruptedException {
        while (cuentas[cuentaOrigen] < cantidad) {
            //Hace esperar hasta que haya suficientes fondos en la cuenta de origen
            wait();
        }
        //Imprime información sobre el hilo actual en la consola
        System.out.println(Thread.currentThread());
    
        //Se realiza la transferencia y se muestra la informacion de las cuentas.
        cuentas[cuentaOrigen] -= cantidad;
        System.out.printf("\n%10.2f de %d para %d.", cantidad, cuentaOrigen, cuentaDestino);
        cuentas[cuentaDestino] += cantidad;
        System.out.printf(" Saldo total: %10.2f",cuentas[cuentaDestino]);
        //Para notificar a todos los hilos en espera que se ha realizado una transferencia.
        //Permite que los hilos verifiquen nuevamente sus condiciones para determinar si pueden continuar con sus tareas.
        notifyAll();
    }

}

File: TALLER_HILOS_DOTNET_GR01/APLICATIVO/PARTE 3/Sincronizacion_Hilos_Dotnet_Banco/Sincronizacion_Hilos_Dotnet_Banco/ec/edu/monster/modelo/Banco.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sincronizacion_Hilos_Dotnet_Banco.ec.edu.monster.modelo
{
    class Banco
    {
        private double[] cuentas;

        //Constructor inicializa 100 cuentas con saldo de 2000 dolares
        public Banco()
        {
            this.cuentas = new double[100];
            for (int i = 0; i < cuentas.Length; i++)
            {
                cuentas[i] = 2000;
            }
        }

        public void transferencia(int cuentaOrigen, int cuentaDestino, double cantidad)
        {
            while (cuentas[cuentaOrigen] < cantidad)
            {
                return;
            }
            cuentas[cuentaOrigen] -= cantidad;
            Console.WriteLine($"Transferir ${cantidad} de la cuenta {cuentaOrigen} para la cuenta {cuentaDestino}. \r\n");
            cuentas[cuentaDestino] += cantidad;
            Console.WriteLine($"Saldo total : {cuentas[cuentaDestino]}.\r\n");
        }
         //Se obtiene el saldo total de la cuenta
        public double getSaldoTotal()
        {
            double total = 0;
            foreach (double a in cuentas)
            {
                total += a;
            }
            return total;
        }
    }
}
